contracts/schemes/Competition.sol

```diff
@@ -0,0 +1,349 @@
+        require(proposals[_proposalId].votingStartTime < now, "voting period not started yet");
+        if (proposals[_proposalId].snapshotBlock == 0) {
+            proposals[_proposalId].snapshotBlock = block.number;
+            emit SnapshotBlock(_proposalId, block.number);
+        }
+    }
+
+    /**
+    * @dev getOrderedIndexOfSuggestion return the index of specific suggestion in the winners list.
+    * @param _proposalId proposal id
+    * @param _suggestionId suggestion id
+    */
+    function getOrderedIndexOfSuggestion(bytes32 _proposalId, uint256 _suggestionId)
+    public
+    view
+    returns(uint256 index) {
+        uint256[] memory topSuggestionsOrdered = proposals[_proposalId].topSuggestionsOrdered;
+        /** get how many elements are greater than a given element*/
+        for (uint256 i; i < topSuggestionsOrdered.length; i++) {
+            if (suggestions[topSuggestionsOrdered[i]].totalVotes > suggestions[_suggestionId].totalVotes) {
+                index++;
+            }
+        }
+    }
+
+    /**
+    * @dev refreshTopSuggestions this function maintain a winners list array.
+    * @param _proposalId proposal id
+    * @param _suggestionId suggestion id
+    */
+    function refreshTopSuggestions(bytes32 _proposalId, uint256 _suggestionId) private {
+        uint256[] storage topSuggestionsOrdered = proposals[_proposalId].topSuggestionsOrdered;
+        if (topSuggestionsOrdered.length < proposals[_proposalId].numberOfWinners) {
+            topSuggestionsOrdered.push(_suggestionId);
+        } else {
+         /** get the index of the smallest element **/
+            uint256 smallest = 0;
+            for (uint256 i; i < proposals[_proposalId].numberOfWinners; i++) {
+                if (suggestions[topSuggestionsOrdered[i]].totalVotes <
+                    suggestions[topSuggestionsOrdered[smallest]].totalVotes) {
+                    smallest = i;
+                }
+            }
+
+            if (suggestions[topSuggestionsOrdered[smallest]].totalVotes < suggestions[_suggestionId].totalVotes) {
+                topSuggestionsOrdered[smallest] = _suggestionId;
+            }
+        }
+    }
+
+    /**
+    * @dev redeem a winning suggestion reward
+    * @param _suggestionId suggestionId
+    * @param _beneficiary - the reward beneficiary
+    */
+    function _redeem(uint256 _suggestionId, address payable _beneficiary) private {
+        bytes32 proposalId = suggestions[_suggestionId].proposalId;
+        require(proposalId != bytes32(0), "suggestion not exist");
+        Proposal storage proposal = proposals[proposalId];
+        // solhint-disable-next-line not-rely-on-time
+        require(proposal.endTime < now, "competition is still on");
+        uint256 amount;
+        //check if there is a win
+        for (uint256 i = 0; i < proposal.topSuggestionsOrdered.length; i++) {
+            if (suggestions[proposal.topSuggestionsOrdered[i]].suggester != address(0)) {
```

Comment:
I do not see at all why this iteration is needed at all - can you add  a comment?

Reply:
(seems to me the for loop can be simply removed)